### DISCUSSION TITLE
Query a DAG to discover node(s) which need to be removed to have an isolate with K elements

### DISCUSSION THREAD

**Discussion by AlexTereshenkov** [networkx/networkx#7140] [General]
I use networkx extensively to query a [dependency graph](https://en.wikipedia.org/wiki/Dependency_graph) of a codebase represented as a DAG with nodes representing files and edges representing dependency between them. It's an incredible tool that provides enormously helpful algorithms to gain insights about the dependency graph and how it can be improved.

One of the experimental ideas I had recently is about [isolates](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.isolate.isolates.html#networkx.algorithms.isolate.isolates). It may be interesting to see if it's possible to break connections in the graph to produce isolates, i.e. independent sections of the graph.

With this graph

```mermaid
  graph TD; 
 a --> b;
 a --> c;
 c --> d;
 c --> e;
 c --> f;
```

Question: connection between what nodes should be broken to produce an isolate with at least 4 nodes?
Answer: `a --> c` (if `a` no longer depends on `c`, then `c, d, e, f` are in an isolate, not connected to the rest of the graph).

I wonder what would be a smart way to query the graph to get this question answered? I can't progress anywhere passed the point of breaking connections between all non-leaf nodes, iteratively, and checking if there are any new isolates, but this won't scale in a large codebase.

**Comment by dschult** [comment]
Here's a hack at an algorithm: (it's a fairly specific problem so there probably isn't a published algorithm -- but who knows)
- traverse the DAG adding a node attribute that holds the number of descendants for that node. 
- collect a list of all nodes with `n-1` descendants and with in_degree 1.
- The edges into those nodes would be good edges to try breaking. You'd still need to check the descendants to make sure they don't have other ancestors not in the potential isolate.

These kinds of questions are nice examples to show networkx in action.   :)
